frame/migrations/src/lib.rs

```diff
@@ -0,0 +1,488 @@
+pub type CursorOf<T> =
+	MigrationCursor<<T as Config>::Cursor, <T as frame_system::Config>::BlockNumber>;
+
+/// Convenience alias for [`ActiveCursor`].
+pub type ActiveCursorOf<T> =
+	ActiveCursor<<T as Config>::Cursor, <T as frame_system::Config>::BlockNumber>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type of the runtime.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// All the multi-block migrations to run.
+		///
+		/// Should only be updated in a runtime-upgrade once all the old migrations have completed.
+		/// (Check that `Cursor` is `None`).
+		type Migrations: Get<MigrationsOf<Self>>;
+
+		/// The cursor type that is shared across all migrations.
+		type Cursor: FullCodec + MaxEncodedLen + TypeInfo + Parameter;
+
+		/// The identifier type that is shared across all migrations.
+		type Identifier: FullCodec + MaxEncodedLen + TypeInfo;
+
+		/// Notifications for status updates of a runtime upgrade.
+		///
+		/// Can be used to pause XCM etc.
+		type UpgradeStatusNotify: UpgradeStatusNotify;
+
+		/// The weight to spend each block to execute migrations.
+		type ServiceWeight: Get<Weight>;
+
+		/// Weight information for the calls and functions of this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	/// The currently active migration to run and its cursor.
+	///
+	/// `None` indicates that no migration is running.
+	#[pallet::storage]
+	pub type Cursor<T: Config> = StorageValue<_, CursorOf<T>, OptionQuery>;
+
+	/// Set of all successfully executed migrations.
+	///
+	/// This is used as blacklist, to not re-execute migrations that have not been removed from the
+	/// codebase yet. Governance can regularly clear this out via `clear_historic`.
+	#[pallet::storage]
+	pub type Historic<T: Config> = StorageMap<_, Twox64Concat, T::Identifier, (), OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Runtime upgrade started with `migrations` in the queue.
+		///
+		/// This can be used to design a progress indicator in combination with counting the
+		/// `MigrationCompleted` and `MigrationSkippedHistoric` events.
+		UpgradeStarted { migrations: u32 },
+		/// Runtime upgrade completed with `migrations`.
+		UpgradeCompleted,
+		/// Runtime upgrade failed.
+		///
+		/// This is very bad and will require governance intervention.
+		UpgradeFailed,
+		/// Migration `index` was skipped, since it already executed in the past.
+		MigrationSkippedHistoric { index: u32 },
+		/// Migration `index` made progress.
+		MigrationAdvanced { index: u32, blocks: T::BlockNumber },
+		/// Migration `index` completed.
+		MigrationCompleted { index: u32, blocks: T::BlockNumber },
```

Comment:
I don't see these being used. moreover, what is the `blocks` supposed to be?

Reply:
It is emitted when a migration completed. `blocks` is the number of blocks that it took to complete.  
Will add a comment.